hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -0,0 +1,1504 @@
+        if paramsSet.contains(s) then s.asPath else Value.Lit(Tree.UnitLit(true))
+      // moved when the capture is instantiated
+      val bod = blockBuilder
+        .assign(captureSym, Instantiate(captureCls.sym.asPath, paramsList))
+        .rest(transformed)
+      Lifted(FunDefn(f.owner, f.sym, f.params, bod), captureCls :: newDefns)
+
+  end liftDefnsInFn
+
+  def desugarLambdas(b: Block) =
+    def rewriteOneBlk(b: Block) =
+      var lambdasList: List[(BlockMemberSymbol, Value.Lam)] = Nil
+      val lambdaRewriter = new BlockTransformerNoRec(SymbolSubst()):
+        override def applyValue(v: Value): Value = v match
+          case lam: Value.Lam => 
+            val sym = BlockMemberSymbol("lambda", Nil)
+            lambdasList ::= (sym -> super.applyLam(lam))
+            Value.Ref(sym)
+          case _ => super.applyValue(v)
+      val blk = lambdaRewriter.applyBlock(b)
+      (blk, lambdasList)
+
+    val transformer = new BlockTransformer(SymbolSubst()):
+      override def applyBlock(b: Block): Block =
+        val (newBlk, lambdasList) = rewriteOneBlk(b)
+        val lambdaDefns = lambdasList.map:
+          case (sym, Value.Lam(params, body)) =>
+            FunDefn(None, sym, params :: Nil, body)
+        val ret = lambdaDefns.foldLeft(newBlk):
+          case (acc, defn) => Define(defn, acc)
+        super.applyBlock(ret)
+    transformer.applyBlock(b)
+
+  // top-level
+  def transform(b: Block) =
+    // this is already done once in the lowering, but the handler lowering adds lambdas currently
+    // so we need to desugar them again
+    val blk = desugarLambdas(b)
+
+    val analyzer = UsedVarAnalyzer(blk)
+    val ctx = LifterCtx
+      .withLocals(analyzer.findUsedLocals)
+      .withDefns(analyzer.defnsMap)
+      .withNestedDefns(analyzer.nestedDefns)
+      .withAccesses(analyzer.accessMap)
+      .withInScopes(analyzer.inScopeDefns)
+
+    var unliftable: Set[BlockMemberSymbol] = Set.empty
+    var modules: List[ClsLikeDefn] = List.empty
+    var objects: List[ClsLikeDefn] = List.empty
+
+
+    val walker = new BlockTransformerShallow(SymbolSubst()):
+      override def applyDefn(defn: Defn): Defn = 
+        val (unl, mod, obj) = createMetadata(defn, ctx)
+        unliftable ++= unl
+        modules ++= mod
+        objects ++= obj
+        defn
+
+    walker.applyBlock(b)
+
+    val modLocals = (modules ++ objects).map: c =>
+      analyzer.nestedIn.get(c.sym) match
+        case Some(bms) =>
+          val nestedIn = analyzer.defnsMap(bms)
+          nestedIn match
+            case cls: ClsLikeDefn => S(c.sym -> TermSymbol(syntax.ImmutVal, S(cls.isym), Tree.Ident(c.sym.nme + "$")))
+            case _ => S(c.sym -> VarSymbol(Tree.Ident(c.sym.nme + "$")))
+        case _ => N
+    .collect:
+      case S(v) => v
+    .toMap
+
+    val ctxx = ctx
+      .addIgnored(unliftable)
+      .withModLocals(modLocals)
+
+    val walker1 = new BlockTransformerShallow(SymbolSubst()):
```

Comment:
Same remarks as before regarding possibly merging walkers and also the fact it's unnecessary to name this instance.